### PR TITLE
DEVELOPERS-116: Upgrade RabbitMQ Client From 5.9.0 To 5.14.1

### PR DESCRIPTION
**Pull request by sharyanrtr**
https://renttherunway.jira.com/browse/DEVELOPERS-116

Main task here is to upgrade the RabbitMQ java client to take advantage of some bug fixes that may solve certain issues we are seeing in cloud environments.

In particular, of note this bugfix is of interest to us: https://github.com/rabbitmq/rabbitmq-java-client/pull/667.

Should overlap with https://github.com/RentTheRunway/conduit/pull/62.

Other improvements while here;

* Use Github Actions For PR checks (remove Jenkinsfile/Travis)
* Migrate from JUnit4 -> JUnit5
* Remove need for qpid with testcontainers
* Simulate network outages with toxiproxy, and auto-reconnect
* Test logging to diagnose issues
* Structure unit/integration tests into own packages

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -96,48 +95,123 @@
...
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
 
     <dependencies>
+        <!-- Core library dependencies -->
+        <dependency>
+            <groupId>com.rabbitmq</groupId>
+            <artifactId>amqp-client</artifactId>
+            <version>5.14.1</version>
```
Comment (sharyanrtr) [inline pom.xml:141]: Main crux of this change. Everything was aiming to get better testing coverage for the upgrade.

File: src/test/java/io/rtr/conduit/integration/ShutdownHandlerIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.core.Is.is;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+@Testcontainers
+public class ShutdownHandlerIntegrationTest {
```
Comment (sharyanrtr) [inline src/test/java/io/rtr/conduit/integration/ShutdownHandlerIntegrationTest.java:26]: This test was extracted in essence from AMQPIntegrationIT.

The test aims to verify the behaviour of Conduit and the RabbitMQ client when an temporary outage happens. We are using https://github.com/Shopify/toxiproxy and testcontainer's customisation for it https://www.testcontainers.org/modules/toxiproxy/ to simulate a network outage.

Comment (sharyanrtr) [reply src/test/java/io/rtr/conduit/integration/ShutdownHandlerIntegrationTest.java:26]: Not the end of this kind of testing, this harness should give us a platform to test temporary outages or what is happening with dynamic/non-dynamic queues disappearing, etc.

File: .github/workflows/pr-validation.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      (github.actor != 'dependabot[bot]' && github.event_name != 'pull_request_target') ||
+      (github.actor == 'dependabot[bot]' && github.event_name == 'pull_request_target')
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v2
+        if: github.actor != 'dependabot[bot]'
+
+      # See: https://hugo.alliau.me/2021/05/04/migration-to-github-native-dependabot-solutions-for-auto-merge-and-action-secrets/#share-your-secrets-with-dependabot
+      - name: Checkout Repository [Dependabot]
```
Comment (FalconerTC) [inline .github/workflows/pr-validation.yaml:28]: Interesting. We could also override `GITHUB_TOKEN` with a RW one from secrets I think

**Review by MatthewRyanRead — APPROVED** [review]
really excellent

**Comment by FalconerTC** [comment]
lgtm but I didn't review the Java

**Review by torypeterschild — APPROVED** [review]

File: .github/workflows/pr-validation.yaml (hunk 3 above)
Comment (sharyanrtr) [reply .github/workflows/pr-validation.yaml:28]: Shamelessly copied from https://github.com/RentTheRunway/dw_parent_pom/blob/f324deddb415496c01c81dbfe3c2151d2e1fee14/.github/workflows/pr-validation.yml#L39

One thing we probably could do with updating is the release jobs, to follow what we have for dw_parent_pom and such.

Hence my question about the open-source'ability of this. We might not be able to straight follow what we have for internal core libraries.

**Review by smeiring — APPROVED** [review]

File: src/test/java/io/rtr/conduit/util/LoggingAmqpCallbackHandler.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import com.rabbitmq.client.ShutdownSignalException;
+import io.rtr.conduit.amqp.AMQPConsumerCallback;
+import io.rtr.conduit.amqp.AMQPMessageBundle;
+import io.rtr.conduit.amqp.ActionResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+
+public class LoggingAmqpCallbackHandler implements AMQPConsumerCallback {
```
Comment (aschutz-rtr) [inline src/test/java/io/rtr/conduit/util/LoggingAmqpCallbackHandler.java:12]: That kinda voids https://renttherunway.jira.com/browse/DEVELOPERS-123 (Standard implementation for AMQP Callbacks) partially, thank you! 👍

**Review by aschutz-rtr — APPROVED** [review]
🚀

**Comment by aschutz-rtr** [comment]
Seeing that most churn resulted from code style changes, I'm wondering what the appetite is for pre-commit hooks that standardise such things?

**Comment by smeiring** [comment]
> 

Using prettier in the js codebases. 
any suggestions @aschutz-rtr ? Something we can definitely advocate for

**Comment by MatthewRyanRead** [comment]
We use Google Java Format in several of our repos.  Really simple to hook up as Maven and IntelliJ plugins: https://renttherunway.jira.com/wiki/spaces/MEM/pages/1228441068/Feb+2020+Using+Google+s+Java+Style+Guide.  Can also be hooked into checkstyle, a pre-commit hook, whatever you want, but that page goes into why my team only uses those plugins.

**Comment by aschutz-rtr** [comment]
In EPIC we've used pre-commit hooks based on https://pre-commit.com/ and code-style based on `pretty-format-java`, `pretty-format-kotlin` and a bunch of other standardizations for yaml, json etc.
Never got to `ktlint` / `detekt` to work properly .

I think Royalties have gotten that to work, no? @bronaghdeane &  @owen-obrien ?

**Comment by owen-obrien** [comment]
> 

Not sure what the question is, ktlint is ran as part of the pretty-format-kotlin hook and seems to behave as expected

`KTLint...................................................................Failed
- hook id: pretty-format-kotlin
- exit code: 1
- files were modified by this hook`

**Comment by sharyanrtr** [comment]
Worth having this discussion in a more public forum like #guild-backend? @owen-obrien @aschutz-rtr @MatthewRyanRead @smeiring. Would agree something consistent across all repos would be great, and GHAs to enforce will take a fair amount of churn on PRs away.

**Pull request merged**